src/state_migration/nv17/migration.rs

```diff
@@ -0,0 +1,191 @@
+};
+
+impl<BS: Blockstore + Clone + Send + Sync> StateMigration<BS> {
+    pub fn add_nv17_migrations(
+        &mut self,
+        store: BS,
+        actors_in: &mut StateTree<BS>,
+        new_manifest: &Manifest,
+        prior_epoch: ChainEpoch,
+        chain_config: &ChainConfig,
+    ) -> anyhow::Result<()> {
+        let system_actor = actors_in
+            .get_actor(&Address::new_id(0))?
+            .ok_or_else(|| anyhow!("system actor not found"))?;
+
+        let system_actor_state: SystemStateOld = store
+            .get_cbor(&system_actor.state)?
+            .ok_or_else(|| anyhow!("system actor state not found"))?;
+        let current_manifest_data = system_actor_state.builtin_actors;
+
+        let current_manifest = Manifest::load_with_actors(&store, &current_manifest_data, 1)?;
+
+        let verifreg_actor_v8 = actors_in
+            .get_actor(&fil_actors_shared::v8::VERIFIED_REGISTRY_ACTOR_ADDR.into())?
+            .context("Failed to load verifreg actor v8")?;
+
+        let verifreg_state_v8: fil_actor_verifreg_state::v8::State = store
+            .get_cbor(&verifreg_actor_v8.state)?
+            .context("Failed to load verifreg state v8")?;
+
+        let market_actor_v8 = actors_in
+            .get_actor(&fil_actors_shared::v8::STORAGE_MARKET_ACTOR_ADDR.into())?
+            .context("Failed to load market actor v8")?;
+
+        let market_state_v8: fil_actor_market_state::v8::State = store
+            .get_cbor(&market_actor_v8.state)?
+            .context("Failed to load market state v8")?;
+
+        let init_actor_v8 = actors_in
+            .get_actor(&fil_actors_shared::v8::INIT_ACTOR_ADDR.into())?
+            .context("Failed to load init actor v8")?;
+
+        let init_state_v8: fil_actor_init_state::v8::State = store
+            .get_cbor(&init_actor_v8.state)?
+            .context("Failed to load init state v8")?;
+
+        let (pending_verified_deals, pending_verified_deal_size) =
+            get_pending_verified_deals_and_total_size(&store, &market_state_v8)?;
+
+        current_manifest.builtin_actors().for_each(|(name, code)| {
+            if name == MARKET_ACTOR_NAME || name == VERIFREG_ACTOR_NAME {
+                self.add_migrator(*code, Arc::new(DeferredMigrator))
+            } else {
+                let new_code = new_manifest.code_by_name(name).unwrap();
+                self.add_migrator(*code, nil_migrator(*new_code));
+            }
+        });
+
+        //https://github.com/filecoin-project/go-state-types/blob/master/builtin/v9/migration/top.go#LL176C2-L176C38
+        self.add_migrator(
+            *current_manifest.system_code(),
+            system::system_migrator(new_manifest),
+        );
+
+        // Sets empty datacap actor to migrate from
+        let datacap_code = new_manifest.code_by_name(DATACAP_ACTOR_NAME)?;
+        actors_in.set_actor(
+            &fil_actors_shared::v9::builtin::DATACAP_TOKEN_ACTOR_ADDR.into(),
```

Comment:
I don't quite get it. Aren't we migrating from `v8`?

Reply:
Yeah, only because `datacap` is not defined under v8, see https://github.com/ChainSafe/fil-actor-states/blob/main/fil_actors_shared/src/v8/builtin/singletons.rs#L7

Reply:
Let's note it down in the comment. Additionally, it'd be fantastic to fix it in the `fil-actor-states`.

Reply:
Since address id is identical across versions, maybe we should switch to version-less address in `fil-actor-states`, like https://github.com/ChainSafe/fil-actor-states/blob/main/fil_actor_interface/src/builtin/cron/mod.rs#L13

Reply:
Sounds good to me.

Reply:
@LesnyRumcajs Actually `datacap` is a new actor introduced in v9, see https://github.com/filecoin-project/go-state-types/blob/1e6cf0d47cdda75383ef036fc2725d1cf51dbde8/manifest/manifest.go#L50

Reply:
Shouldn't then it be handled like Ethereum ones in NV18 migration? That is to say, in post migration (if I recall correctly)?

Reply:
Yeah but the Go code handles this specially by creating an empty actor to migrate from

Reply:
Is there any difference in the end? I mean, if we have a cleaner way of doing something, it's okay not to do exactly the same as Lotus does.

Reply:
I can have a try

Reply:
Switched to a post migrator with comments

Reply:
Thank you!